keps/sig-api-machinery/3396-websocket-protocol-enhancements/README.md

```diff
@@ -0,0 +1,869 @@
+bogged down.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Protocol changes
+
+Currently, API server supports [binary and base64 WebSocket protocols](https://github.com/kubernetes/kubernetes/blob/3ddd0f45aa91e2f30c70734b175631bec5b5825a/staging/src/k8s.io/apiserver/pkg/util/wsstream/conn.go#L34-L64). 
+
+Binary protocol: channel id `0xFF` can be used as a control channel. First byte of the message contents can be used
+to encode the command and the rest of the message can be used to encode the payload for the command.
+
+Binary protocol control commands:
+
+- `0x00`: half-close the channel. Next byte is the channel id.
+- `0x01`: reset the channel. Next byte is the channel id.
+
+Base64-based protocol: can use character `c` (for `control`) as the channel id for the control channel.
+
+Base64-based protocol control commands:
+
+- `h`: half-close the channel. Next character is the channel id.
+- `r`: reset the channel. Next character is the channel id.
+
+### Protocol negotiation
+
+If we have a new protocol, then older servers will not have support for it. In that case a newer client will try
+v5 but an older server will not accept that. To be backwards compatible, ideally client should use the intended
+mechanism that HTTP `Upgrade` provides - send a list of protocols. I.e. client would send a header with
+something like `Upgrade: v5.channel.k8s.io, SPDY/3.1` and server would pick a protocol it supports,
+taking into account client's preference (the list is ordered).
+
+The difficulty with the above is that both Gorilla WebSocket and SPDY libraries are not built to allow to
+use them this way. Both encapsulate negotiation, and you cannot compose them to get the above behavior. 
+Refactoring them is not an option since SPDY is dead and nobody will spend time on the library and Gorilla
+WebSocket is [looking for maintainers](https://github.com/gorilla/websocket/issues/370) and hence
+a PR is unlikely to get merged quickly even if at all (since this is a big API addition/change).
```

Comment:
Perhaps the Kubernetes API could provide a way for the cluster to expose what protocols it can negotiate?

Eg, `GET /api/v1alpha1/protocol-availability` (or something similar)

Reply:
This wouldn't reduce the number of API calls compared to what is being suggesting in this PR right now. The current suggestion will do 1 more call temporarily, and only when client is newer than the server. Eventually old clients will come out of use and clients will not be affected anymore.